Certificate serial numbers are ocassionally invalid ASN.1 DER integers

I believe this is the culprit:

https://github.com/relaycorp/relaynet-core-js/blob/309ce368ac17434f9f9586a2c9cb9b6a1a73be5b/src/lib/crypto_wrappers/x509/Certificate.ts#L65-L68

It happens very rarely -- maybe 2% of the time -- and when it does, Bouncy Castle would reject it. For example, if it happens when registering with the public gateway, here's the traceback:

```
tech.relaycorp.relaynet.bindings.pdc.ServerBindingException: The server returned a malformed registration

	at tech.relaycorp.poweb.PoWebClient.registerNode(PoWebClient.kt:129)
	at tech.relaycorp.poweb.PoWebClient$registerNode$1.invokeSuspend(PoWebClient.kt)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:274)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:84)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at tech.relaycorp.gateway.RealTest.test_connection(RealTest.kt:20)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:688)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestTemplateMethod(TimeoutExtension.java:92)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:210)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:206)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:131)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:65)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$DefaultDynamicTestExecutor.execute(NodeTestTask.java:212)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$DefaultDynamicTestExecutor.execute(NodeTestTask.java:192)
	at org.junit.jupiter.engine.descriptor.TestTemplateTestDescriptor.execute(TestTemplateTestDescriptor.java:139)
	at org.junit.jupiter.engine.descriptor.TestTemplateTestDescriptor.lambda$execute$2(TestTemplateTestDescriptor.java:107)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.util.stream.IntPipeline$4$1.accept(IntPipeline.java:250)
	at java.util.stream.Streams$RangeIntSpliterator.forEachRemaining(Streams.java:110)
	at java.util.Spliterator$OfInt.forEachRemaining(Spliterator.java:693)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:485)
	at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:272)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:485)
	at org.junit.jupiter.engine.descriptor.TestTemplateTestDescriptor.execute(TestTemplateTestDescriptor.java:107)
	at org.junit.jupiter.engine.descriptor.TestTemplateTestDescriptor.execute(TestTemplateTestDescriptor.java:42)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:230)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:58)
Caused by: tech.relaycorp.relaynet.messages.InvalidMessageException: Node registration contains invalid node certificate
	at tech.relaycorp.relaynet.messages.control.PrivateNodeRegistration$Companion.deserialize(PrivateNodeRegistration.kt:56)
	at tech.relaycorp.poweb.PoWebClient.registerNode(PoWebClient.kt:127)
	... 112 more
Caused by: tech.relaycorp.relaynet.wrappers.x509.CertificateException: Value should be a DER-encoded, X.509 v3 certificate
	at tech.relaycorp.relaynet.wrappers.x509.Certificate$Companion.deserialize(Certificate.kt:128)
	at tech.relaycorp.relaynet.messages.control.PrivateNodeRegistration$Companion.deserializeCertificate(PrivateNodeRegistration.kt:75)
	at tech.relaycorp.relaynet.messages.control.PrivateNodeRegistration$Companion.deserialize(PrivateNodeRegistration.kt:54)
	... 113 more
Caused by: org.bouncycastle.asn1.ASN1Exception: corrupted stream detected
	at org.bouncycastle.asn1.ASN1InputStream.readObject(Unknown Source)
	at org.bouncycastle.asn1.ASN1InputStream.readVector(Unknown Source)
	at org.bouncycastle.asn1.ASN1InputStream.buildObject(Unknown Source)
	at org.bouncycastle.asn1.ASN1InputStream.readObject(Unknown Source)
	at org.bouncycastle.asn1.ASN1InputStream.readVector(Unknown Source)
	at org.bouncycastle.asn1.ASN1InputStream.buildObject(Unknown Source)
	at org.bouncycastle.asn1.ASN1InputStream.readObject(Unknown Source)
	at org.bouncycastle.asn1.ASN1Primitive.fromByteArray(Unknown Source)
	at org.bouncycastle.cert.CertUtils.parseNonEmptyASN1(Unknown Source)
	at org.bouncycastle.cert.X509CertificateHolder.parseBytes(Unknown Source)
	at org.bouncycastle.cert.X509CertificateHolder.<init>(Unknown Source)
	at tech.relaycorp.relaynet.wrappers.x509.Certificate$Companion.deserialize(Certificate.kt:126)
	... 115 more
Caused by: java.lang.IllegalArgumentException: malformed integer
	at org.bouncycastle.asn1.ASN1Integer.<init>(Unknown Source)
	at org.bouncycastle.asn1.ASN1InputStream.createPrimitiveDERObject(Unknown Source)
	at org.bouncycastle.asn1.ASN1InputStream.buildObject(Unknown Source)
	... 127 more
```

--- a/src/lib/crypto_wrappers/x509/Certificate.spec.ts
+++ b/src/lib/crypto_wrappers/x509/Certificate.spec.ts
@@ -120,7 +120,7 @@ describe('issue()', () => {
         anySignFlipped = true;
       } else {
         expect(serialNumberSerialized).toHaveLength(8);
-        expect(serialNumberSerialized[0]).toBeGreaterThan(0);
+        expect(serialNumberSerialized[0]).toBeGreaterThanOrEqual(0);
         expect(serialNumberSerialized[0]).toBeLessThanOrEqual(127);
       }
     }

--- a/src/lib/crypto_wrappers/x509/Certificate.spec.ts
+++ b/src/lib/crypto_wrappers/x509/Certificate.spec.ts
@@ -10,7 +10,7 @@ import {
   sha256Hex,
 } from '../../_test_utils';
 import * as oids from '../../oids';
-import * as utils from '../_utils';
+import { derDeserialize, getPkijsCrypto } from '../_utils';
 import {
   derSerializePublicKey,
   generateRSAKeyPair,
@@ -22,7 +22,7 @@ import CertificateError from './CertificateError';
 const futureDate = new Date();
 futureDate.setDate(futureDate.getDate() + 1);
 
-const pkijsCrypto = utils.getPkijsCrypto();
+const pkijsCrypto = getPkijsCrypto();
 
 afterEach(() => {
   jest.restoreAllMocks();
@@ -104,17 +104,28 @@ describe('issue()', () => {
     );
   });
 
-  test('should generate a serial number', async () => {
-    const generateRandom64BitValueSpy = jest.spyOn(utils, 'generateRandom64BitValue');
-    const cert = await Certificate.issue({
-      ...baseCertificateOptions,
-      issuerPrivateKey: keyPair.privateKey,
-      subjectPublicKey: keyPair.publicKey,
-    });
+  test('should generate a positive serial number', async () => {
+    let anySignFlipped = false;
+    for (let index = 0; index < 10; index++) {
+      const cert = await Certificate.issue({
+        ...baseCertificateOptions,
+        issuerPrivateKey: keyPair.privateKey,
+        subjectPublicKey: keyPair.publicKey,
+      });
+      const serialNumberSerialized = new Uint8Array(
+        cert.pkijsCertificate.serialNumber.valueBlock.valueHex,
+      );
+      if (serialNumberSerialized.length === 9) {
+        expect(serialNumberSerialized[0]).toEqual(0);
+        anySignFlipped = true;
+      } else {
+        expect(serialNumberSerialized).toHaveLength(8);
+        expect(serialNumberSerialized[0]).toBeGreaterThan(0);
+        expect(serialNumberSerialized[0]).toBeLessThanOrEqual(127);
+      }
+    }
 
-    expect(generateRandom64BitValueSpy).toBeCalledTimes(1);
-    const generatedNumber = generateRandom64BitValueSpy.mock.results[0].value;
-    expect(cert.pkijsCertificate.serialNumber.valueBlock.valueHex).toEqual(generatedNumber);
+    expect(anySignFlipped).toBeTrue();
   });
 
   test('should create a certificate valid from now by default', async () => {
@@ -437,7 +448,7 @@ describe('issue()', () => {
       expect(matchingExtensions).toHaveLength(1);
       const akiExtension = matchingExtensions[0];
       expect(akiExtension.critical).toBe(false);
-      const akiExtensionAsn1 = utils.derDeserialize(akiExtension.extnValue.valueBlock.valueHex);
+      const akiExtensionAsn1 = derDeserialize(akiExtension.extnValue.valueBlock.valueHex);
       const akiExtensionRestored = new pkijs.AuthorityKeyIdentifier({
         schema: akiExtensionAsn1,
       });
@@ -467,7 +478,7 @@ describe('issue()', () => {
       expect(matchingExtensions).toHaveLength(1);
       const akiExtension = matchingExtensions[0];
       expect(akiExtension.critical).toBe(false);
-      const akiExtensionAsn1 = utils.derDeserialize(akiExtension.extnValue.valueBlock.valueHex);
+      const akiExtensionAsn1 = derDeserialize(akiExtension.extnValue.valueBlock.valueHex);
       const akiExtensionRestored = new pkijs.AuthorityKeyIdentifier({
         schema: akiExtensionAsn1,
       });
@@ -500,7 +511,7 @@ describe('issue()', () => {
     expect(matchingExtensions).toHaveLength(1);
     const skiExtension = matchingExtensions[0];
     expect(skiExtension.critical).toBe(false);
-    const skiExtensionAsn1 = utils.derDeserialize(skiExtension.extnValue.valueBlock.valueHex);
+    const skiExtensionAsn1 = derDeserialize(skiExtension.extnValue.valueBlock.valueHex);
     expect(skiExtensionAsn1).toBeInstanceOf(asn1js.OctetString);
     // @ts-ignore
     const keyIdBuffer = Buffer.from(skiExtensionAsn1.valueBlock.valueHex);
@@ -513,7 +524,7 @@ test('serialize() should return a DER-encoded buffer', async () => {
 
   const certDer = cert.serialize();
 
-  const asn1Value = utils.derDeserialize(certDer);
+  const asn1Value = derDeserialize(certDer);
   const pkijsCert = new pkijs.Certificate({ schema: asn1Value });
 
   const subjectDnAttributes = pkijsCert.subject.typesAndValues;
@@ -542,13 +553,12 @@ test('expiryDate should return the expiry date', async () => {
 });
 
 test('getSerialNumber() should return the serial number as a buffer', async () => {
-  const generateRandom64BitValueSpy = jest.spyOn(utils, 'generateRandom64BitValue');
   const cert = await generateStubCert();
 
   const serialNumberBuffer = cert.getSerialNumber();
-  const expectedSerialNumberHex = generateRandom64BitValueSpy.mock.results[0].value;
-  expect(serialNumberBuffer).toEqual(Buffer.from(expectedSerialNumberHex));
-  expect(serialNumberBuffer.equals(Buffer.from(expectedSerialNumberHex))).toBeTrue();
+  expect(serialNumberBuffer).toEqual(
+    Buffer.from(cert.pkijsCertificate.serialNumber.valueBlock.valueHex),
+  );
 });
 
 test('getSerialNumberHex() should return the hex representation of serial number', async () => {
@@ -884,7 +894,7 @@ function getBasicConstraintsExtension(cert: Certificate): pkijs.BasicConstraints
   const extensions = cert.pkijsCertificate.extensions as ReadonlyArray<pkijs.Extension>;
   const matchingExtensions = extensions.filter((e) => e.extnID === oids.BASIC_CONSTRAINTS);
   const extension = matchingExtensions[0];
-  const basicConstraintsAsn1 = utils.derDeserialize(extension.extnValue.valueBlock.valueHex);
+  const basicConstraintsAsn1 = derDeserialize(extension.extnValue.valueBlock.valueHex);
   return new pkijs.BasicConstraints({ schema: basicConstraintsAsn1 });
 }
 

--- a/src/lib/crypto_wrappers/x509/Certificate.ts
+++ b/src/lib/crypto_wrappers/x509/Certificate.ts
@@ -1,4 +1,4 @@
-import * as asn1js from 'asn1js';
+import { BmpString, Integer, LocalBaseBlock, OctetString } from 'asn1js';
 import * as pkijs from 'pkijs';
 import { makeDateWithSecondPrecision } from '../../_utils';
 
@@ -62,17 +62,13 @@ export default class Certificate {
     const issuerPublicKey = options.issuerCertificate
       ? await options.issuerCertificate.pkijsCertificate.getPublicKey()
       : options.subjectPublicKey;
-    const serialNumberBlock = new asn1js.Integer({
-      // @ts-ignore
-      valueHex: generateRandom64BitValue(),
-    });
     const pkijsCert = new pkijs.Certificate({
       extensions: [
         makeBasicConstraintsExtension(options.isCA === true, options.pathLenConstraint ?? 0),
         await makeAuthorityKeyIdExtension(issuerPublicKey),
         await makeSubjectKeyIdExtension(options.subjectPublicKey),
       ],
-      serialNumber: serialNumberBlock,
+      serialNumber: generatePositiveASN1Integer(),
       version: 2, // 2 = v3
     });
 
@@ -84,7 +80,7 @@ export default class Certificate {
     pkijsCert.subject.typesAndValues.push(
       new pkijs.AttributeTypeAndValue({
         type: oids.COMMON_NAME,
-        value: new asn1js.BmpString({ value: options.commonName }),
+        value: new BmpString({ value: options.commonName }),
       }),
     );
 
@@ -270,6 +266,23 @@ export default class Certificate {
   }
 }
 
+function generatePositiveASN1Integer(): Integer {
+  const signedInteger = generateRandom64BitValue();
+  const signedIntegerView = new Uint8Array(signedInteger);
+
+  let unsignedInteger = signedIntegerView;
+  if (127 < signedIntegerView[0]) {
+    // The integer is negative, so let's flip the sign by prepending a 0x00 octet. See:
+    // https://docs.microsoft.com/en-us/windows/win32/seccertenroll/about-integer
+    unsignedInteger = new Uint8Array(signedIntegerView.byteLength + 1);
+    unsignedInteger.set(signedIntegerView, 1);
+  }
+
+  return new Integer({
+    valueHex: unsignedInteger,
+  } as any);
+}
+
 //region Extensions
 
 function makeBasicConstraintsExtension(cA: boolean, pathLenConstraint: number): pkijs.Extension {
@@ -288,7 +301,7 @@ function makeBasicConstraintsExtension(cA: boolean, pathLenConstraint: number): 
 
 async function makeAuthorityKeyIdExtension(publicKey: CryptoKey): Promise<pkijs.Extension> {
   const keyDigest = await getPublicKeyDigest(publicKey);
-  const keyIdEncoded = new asn1js.OctetString({ valueHex: keyDigest });
+  const keyIdEncoded = new OctetString({ valueHex: keyDigest });
   return new pkijs.Extension({
     extnID: oids.AUTHORITY_KEY,
     extnValue: new pkijs.AuthorityKeyIdentifier({ keyIdentifier: keyIdEncoded })
@@ -301,7 +314,7 @@ async function makeSubjectKeyIdExtension(publicKey: CryptoKey): Promise<pkijs.Ex
   const keyDigest = await getPublicKeyDigest(publicKey);
   return new pkijs.Extension({
     extnID: oids.SUBJECT_KEY,
-    extnValue: new asn1js.OctetString({ valueHex: keyDigest }).toBER(false),
+    extnValue: new OctetString({ valueHex: keyDigest }).toBER(false),
   });
 }
 
@@ -329,7 +342,7 @@ interface Asn1jsSerializable {
   readonly toBER: (sizeOnly?: boolean) => ArrayBuffer;
 }
 
-function cloneAsn1jsValue(value: Asn1jsSerializable): asn1js.LocalBaseBlock {
+function cloneAsn1jsValue(value: Asn1jsSerializable): LocalBaseBlock {
   const valueSerialized = value.toBER(false);
   return derDeserialize(valueSerialized);
 }

--- a/src/lib/crypto_wrappers/x509/Certificate.ts
+++ b/src/lib/crypto_wrappers/x509/Certificate.ts
@@ -274,7 +274,7 @@ function generatePositiveASN1Integer(): Integer {
     // The integer is negative, so let's flip the sign by prepending a 0x00 octet. See:
     // https://docs.microsoft.com/en-us/windows/win32/seccertenroll/about-integer
     unsignedInteger = new Uint8Array(signedInteger.byteLength + 1);
-    unsignedInteger.set(signedInteger, 1);
+    unsignedInteger.set(signedInteger, 1); // Skip the first octet, leaving it as 0x00
   }
 
   return new Integer({

--- a/src/lib/crypto_wrappers/x509/Certificate.ts
+++ b/src/lib/crypto_wrappers/x509/Certificate.ts
@@ -267,15 +267,14 @@ export default class Certificate {
 }
 
 function generatePositiveASN1Integer(): Integer {
-  const signedInteger = generateRandom64BitValue();
-  const signedIntegerView = new Uint8Array(signedInteger);
+  const signedInteger = new Uint8Array(generateRandom64BitValue());
 
-  let unsignedInteger = signedIntegerView;
-  if (127 < signedIntegerView[0]) {
+  let unsignedInteger = signedInteger;
+  if (127 < signedInteger[0]) {
     // The integer is negative, so let's flip the sign by prepending a 0x00 octet. See:
     // https://docs.microsoft.com/en-us/windows/win32/seccertenroll/about-integer
-    unsignedInteger = new Uint8Array(signedIntegerView.byteLength + 1);
-    unsignedInteger.set(signedIntegerView, 1);
+    unsignedInteger = new Uint8Array(signedInteger.byteLength + 1);
+    unsignedInteger.set(signedInteger, 1);
   }
 
   return new Integer({

